sql: TestSavepoints failed

sql.TestSavepoints [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3443717&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3443717&tab=artifacts#/) on master @ [a79a08268cbaad620d2e2dc0c3e19dd670795a1a](https://github.com/cockroachdb/cockroach/commits/a79a08268cbaad620d2e2dc0c3e19dd670795a1a):


```
=== RUN   TestSavepoints
    test_log_scope.go:76: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSavepoints887684322
    test_log_scope.go:77: use -show-logs to present logs inline
=== CONT  TestSavepoints
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSavepoints887684322
--- FAIL: TestSavepoints (1.14s)
=== RUN   TestSavepoints/rollback_after_lock_timeout
    conn_executor_savepoints_test.go:107: 7: reading progress: pq: canceling statement due to lock timeout on row (conn,n)=('default',3) in progress@primary
    datadriven.go:243: 
        testdata/savepoints:265: SET lock_timeout = '1ms'
        BEGIN
        SAVEPOINT foo
        SELECT * FROM t WHERE x = 1 FOR UPDATE
        ROLLBACK TO SAVEPOINT foo
        SELECT * FROM t WHERE x = 2 FOR UPDATE
        COMMIT
        SET lock_timeout = 0
        output didn't match expected:
        @@ -9,9 +9,9 @@
         5: ROLLBACK TO SAVEPOINT foo -- 0 rows
         -- Aborted     -> Open        ###.....  foo
         6: SELECT * FROM t WHERE x = 2 FOR UPDATE -- 0 rows
         -- Open        -> Open        ###..#..  foo
         7: COMMIT -- 0 rows
        --- Open        -> NoTxn       ###..##.  (none)
        +-- Open        -> NoTxn       ........  (none)
         8: SET lock_timeout = 0 -- 0 rows
         -- NoTxn       -> NoTxn       ###..###  (none)
         
    --- FAIL: TestSavepoints/rollback_after_lock_timeout (0.10s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestSavepoints PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries andreimatei
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSavepoints.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/conn_executor_savepoints_test.go
+++ b/pkg/sql/conn_executor_savepoints_test.go
@@ -36,10 +36,26 @@ func TestSavepoints(t *testing.T) {
 	datadriven.Walk(t, "testdata/savepoints", func(t *testing.T, path string) {
 
 		params := base.TestServerArgs{}
-		s, origConn, _ := serverutils.StartServer(t, params)
+		s, db, _ := serverutils.StartServer(t, params)
 		defer s.Stopper().Stop(ctx)
 
-		if _, err := origConn.Exec(`CREATE TABLE progress(
+		sqlConns := make(map[string]*gosql.Conn)
+		getConn := func(name string) *gosql.Conn {
+			conn, ok := sqlConns[name]
+			if ok {
+				return conn
+			}
+			conn, err := db.Conn(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+			sqlConns[name] = conn
+			return conn
+		}
+
+		// progressConn is used to manipulate the progress table.
+		progressConn := getConn("progress")
+		if _, err := progressConn.ExecContext(ctx, `CREATE TABLE progress(
       conn STRING,
     	n INT, 
     	marker BOOL,
@@ -48,33 +64,21 @@ func TestSavepoints(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		const defaultConn = "default"
-		sqlConns := make(map[string]*gosql.DB)
-		sqlConns[defaultConn] = origConn
-
 		datadriven.RunTest(t, path, func(t *testing.T, td *datadriven.TestData) string {
 			switch td.Cmd {
 			case "sql":
 				// Determine which connection to run the SQL statements on. If no
 				// connection specifier is provided, use the "default" conn.
-				connName := defaultConn
-				sqlConn := sqlConns[connName]
+				connName := "default"
 				if td.HasArg("conn") {
 					td.ScanArgs(t, "conn", &connName)
-
-					var ok bool
-					sqlConn, ok = sqlConns[connName]
-					if !ok {
-						sqlConn = serverutils.OpenDBConn(
-							t, s.ServingSQLAddr(), params.UseDatabase, params.Insecure, s.Stopper())
-						sqlConns[connName] = sqlConn
-					}
 				}
+				sqlConn := getConn(connName)
 
 				// Implicitly abort any previously-ongoing txn.
-				_, _ = sqlConn.Exec("ABORT")
+				_, _ = sqlConn.ExecContext(ctx, "ABORT")
 				// Prepare for the next test.
-				if _, err := sqlConn.Exec("DELETE FROM progress WHERE conn = $1", connName); err != nil {
+				if _, err := progressConn.ExecContext(ctx, "DELETE FROM progress WHERE conn = $1", connName); err != nil {
 					td.Fatalf(t, "cleaning up: %v", err)
 				}
 
@@ -104,7 +108,7 @@ func TestSavepoints(t *testing.T) {
 				// updateProgress loads the current set of writes
 				// into the progress bar.
 				updateProgress := func() {
-					rows, err := sqlConn.Query("SELECT n FROM progress WHERE conn = $1", connName)
+					rows, err := progressConn.QueryContext(ctx, "SELECT n FROM progress WHERE conn = $1", connName)
 					if err != nil {
 						t.Logf("%d: reading progress: %v", stepNum, err)
 						// It's OK if we can't read this.
@@ -127,7 +131,7 @@ func TestSavepoints(t *testing.T) {
 				// This is guaranteed to always succeed because SHOW TRANSACTION STATUS
 				// is an observer statement.
 				getTxnStatus := func() string {
-					row := sqlConn.QueryRow("SHOW TRANSACTION STATUS")
+					row := sqlConn.QueryRowContext(ctx, "SHOW TRANSACTION STATUS")
 					var status string
 					if err := row.Scan(&status); err != nil {
 						td.Fatalf(t, "%d: unable to retrieve txn status: %v", stepNum, err)
@@ -137,7 +141,7 @@ func TestSavepoints(t *testing.T) {
 				// showSavepointStatus is like getTxnStatus but retrieves the
 				// savepoint stack.
 				showSavepointStatus := func() {
-					rows, err := sqlConn.Query("SHOW SAVEPOINT STATUS")
+					rows, err := sqlConn.QueryContext(ctx, "SHOW SAVEPOINT STATUS")
 					if err != nil {
 						td.Fatalf(t, "%d: unable to retrieve savepoint status: %v", stepNum, err)
 					}
@@ -184,7 +188,7 @@ func TestSavepoints(t *testing.T) {
 					// Before each statement, mark the progress so far with
 					// a KV write.
 					if isOpenTxn(beforeStatus) {
-						_, err := sqlConn.Exec("INSERT INTO progress(conn, n, marker) VALUES ($1, $2, true)", connName, stepNum)
+						_, err := progressConn.ExecContext(ctx, "INSERT INTO progress(conn, n, marker) VALUES ($1, $2, true)", connName, stepNum)
 						if err != nil {
 							td.Fatalf(t, "%d: before-stmt: %v", stepNum, err)
 						}
@@ -192,7 +196,7 @@ func TestSavepoints(t *testing.T) {
 
 					// Run the statement and report errors/results.
 					fmt.Fprintf(&buf, "%d: %s -- ", stepNum, stmt)
-					execRes, err := sqlConn.Exec(stmt)
+					execRes, err := sqlConn.ExecContext(ctx, stmt)
 					if err != nil {
 						fmt.Fprintf(&buf, "%v\n", err)
 					} else {

